lib/archethic/contracts.ex

```diff
@@ -40,98 +39,157 @@ defmodule Archethic.Contracts do
+
+  # In the case of a trigger interval,
+  # because of the delay between execution and validation,
+  # we override the value returned by library function Time.now()
+  defp trigger_to_execute_opts({:interval, datetime}), do: [time_now: datetime]
+  defp trigger_to_execute_opts(_), do: []
+
+  @doc """
+  Validate any kind of condition.
+  The transaction and datetime depends on the condition.
+  """
+  @spec valid_condition?(
+          :oracle | :transaction | :inherit,
+          Contract.t(),
+          Transaction.t(),
+          DateTime.t()
+        ) :: boolean()
+  def valid_condition?(
+        condition_type,
+        contract = %Contract{version: version, conditions: conditions},
+        transaction = %Transaction{},
+        datetime
+      ) do
+    case Map.get(conditions, condition_type) do
       nil ->
-        :ok
+        true
 
-      inherit_conditions ->
-        constants = %{
-          "previous" => Constants.from_transaction(prev_tx),
-          "next" => Constants.from_transaction(next_tx)
-        }
+      condition ->
+        constants = get_condition_constants(condition_type, contract, transaction, datetime)
+        Interpreter.valid_conditions?(version, condition, constants)
+    end
+  end
 
-        if Interpreter.valid_conditions?(version, inherit_conditions, constants) do
-          :ok
-        else
-          Logger.error("Inherit constraints not respected")
-          {:error, :invalid_inherit_constraints}
-        end
+  defp validate_trigger({:datetime, datetime}, _trigger, trigger_datetime) do
+    if within_drift_tolerance?(trigger_datetime, datetime) do
+      :ok
+    else
+      :invalid_triggers_execution
     end
   end
 
-  defp validate_triggers(triggers, _next_tx, _date) when map_size(triggers) == 0, do: :ok
+  defp validate_trigger({:interval, interval}, {:interval, interval_datetime}, trigger_datetime) do
+    matches_date? =
+      interval
+      |> CronParser.parse!(@extended_mode?)
+      |> CronDateChecker.matches_date?(DateTime.to_naive(interval_datetime))
 
-  defp validate_triggers(triggers, next_tx, date) do
-    if Enum.any?(triggers, fn {trigger_type, _} ->
-         valid_from_trigger?(trigger_type, next_tx, date)
-       end) do
+    if matches_date? && within_drift_tolerance?(trigger_datetime, interval_datetime) do
       :ok
     else
-      Logger.error("Transaction not processed by a valid smart contract trigger")
-      {:error, :invalid_triggers_execution}
+      :invalid_triggers_execution
     end
   end
 
-  defp valid_from_trigger?(
-         {:datetime, datetime},
-         %Transaction{},
-         validation_date = %DateTime{}
-       ) do
-    # Accept time drifing for 10seconds
-    DateTime.diff(validation_date, datetime) >= 0 and
-      DateTime.diff(validation_date, datetime) < 10
+  defp validate_trigger(:transaction, {:transaction, _address}, _) do
+    # maybe check address exist and contain the contract in the recipients?
```

Comment:
indeed this could be interesting. The function `validate_trigger` could return  `{:ok, maybe_tx}` and in case of oracle or transaction trigger fetch the transaction and verify if it's recipient (resolved address in validation stamp) is the previous contract address.